applications/FluidDynamicsApplication/custom_elements/compressible_navier_stokes_explicit.h

```diff
@@ -508,6 +513,550 @@ class CompressibleNavierStokesExplicit : public Element
+{
+    // Get geometry data
+    const auto& r_geom = GetGeometry();
+    const unsigned int NumNodes = r_geom.PointsNumber();
+    Geometry<Node<3>>::ShapeFunctionsGradientsType dNdX_container;
+    r_geom.ShapeFunctionsIntegrationPointsGradients(dNdX_container, GeometryData::IntegrationMethod::GI_GAUSS_1);
+    const auto& r_dNdX = dNdX_container[0];
+
+    // Calculate midpoint magnitudes
+    const double c_v = GetProperties()[SPECIFIC_HEAT];
+    array_1d<double,3> midpoint_grad_temp = ZeroVector(3);
+    for (unsigned int i_node = 0; i_node < NumNodes; ++i_node) {
+        auto& r_node = r_geom[i_node];
+        const auto node_dNdX = row(r_dNdX, i_node);
+        const auto& r_mom = r_node.FastGetSolutionStepValue(MOMENTUM);
+        const double& r_rho = r_node.FastGetSolutionStepValue(DENSITY);
+        const double& r_tot_ener = r_node.FastGetSolutionStepValue(TOTAL_ENERGY);
+        const array_1d<double, 3> vel = r_mom / r_rho;
+        const double temp = (r_tot_ener / r_rho - 0.5 * inner_prod(vel, vel)) / c_v;
+        for (unsigned int d1 = 0; d1 < TDim; ++d1) {
+            midpoint_grad_temp[d1] += node_dNdX(d1) * temp;
+        }
+    }
+
+    return midpoint_grad_temp;
+}
+
+template <unsigned int TDim, unsigned int TNumNodes>
+double CompressibleNavierStokesExplicit<TDim, TNumNodes>::CalculateMidPointSoundVelocity() const
+{
+    // Get geometry data
+    const auto& r_geom = GetGeometry();
+    const unsigned int NumNodes = r_geom.PointsNumber();
+
+    // Calculate midpoint magnitudes
+    double midpoint_rho = 0.0;
+    double midpoint_tot_ener = 0.0;
+    array_1d<double,TDim> midpoint_mom = ZeroVector(TDim);
+    for (unsigned int i_node = 0; i_node < NumNodes; ++i_node) {
+        auto& r_node = r_geom[i_node];
+        const auto& r_mom = r_node.FastGetSolutionStepValue(MOMENTUM);
+        const double& r_rho = r_node.FastGetSolutionStepValue(DENSITY);
+        const double& r_tot_ener = r_node.FastGetSolutionStepValue(TOTAL_ENERGY);
+        midpoint_rho += r_rho;
+        midpoint_tot_ener += r_tot_ener;
+        for (unsigned int d1 = 0; d1 < TDim; ++d1) {
+            midpoint_mom[d1] += r_mom(d1);
+        }
+    }
+    midpoint_rho /= NumNodes;
+    midpoint_mom /= NumNodes;
+    midpoint_tot_ener /= NumNodes;
+
+    // Calculate midpoint speed of sound
+    const auto& r_prop = GetProperties();
+    const double c_v = r_prop.GetValue(SPECIFIC_HEAT);
+    const double gamma = r_prop.GetValue(HEAT_CAPACITY_RATIO);
+    const double temp = (midpoint_tot_ener / midpoint_rho - inner_prod(midpoint_mom, midpoint_mom) / (2 * std::pow(midpoint_rho, 2))) / c_v;
+    double midpoint_c = std::sqrt(gamma * (gamma - 1.0) * c_v * temp);
+    return midpoint_c;
+}
+
+template <unsigned int TDim, unsigned int TNumNodes>
+double CompressibleNavierStokesExplicit<TDim, TNumNodes>::CalculateMidPointVelocityDivergence() const
```

Comment:
Just mentioning that we'll need to change the implementation of this when switching to the alternative interpolation.